### PR TITLE
Update description on rule based index selection

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What is changed, added or deleted? (Required)

Update description on rule based index selection.

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: None
- Other reference link(s): https://github.com/pingcap/tidb/issues/26020

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: choose-index.md [hunk 1]
```diff
@@ -29,19 +29,35 @@ aliases: ['/docs-cn/dev/choose-index/']
...
+### 基于规则选择
 
-Skyline-Pruning 是一个针对索引的启发式过滤规则，评判一个索引的好坏需要从以下三个维度进行衡量：
+#### 前置规则
 
-- 选择该索引读表时，是否需要回表（即该索引生成的计划是 IndexReader 还是 IndexLookupReader）。不用回表的索引在这个维度上优于需要回表的索引。
+TiDB 采用如下的启发式前置规则来选择索引：
 
-- 选择该索引是否能满足一定的顺序。因为索引的读取可以保证某些列集合的顺序，所以满足查询要求顺序的索引在这个维度上优于不满足的索引。
+1. 存在索引满足“唯一性索引全匹配 + 不需要回表”时，直接选择该索引。
```
Comment (qiancai) [inline choose-index.md:40]:
```suggestion
1. 存在索引满足“唯一性索引全匹配 + 不需要回表（即该索引生成的计划是 IndexReader）”时，直接选择该索引。
```

File: choose-index.md [hunk 2]
```diff
@@ -29,19 +29,35 @@ aliases: ['/docs-cn/dev/choose-index/']
...
+
+2. 存在索引满足“唯一性索引 + 需要回表”时，选择满足条件的回表行数最小的索引作为候选索引。
```
Comment (qiancai) [inline choose-index.md:42]:
```suggestion
2. 存在索引满足“唯一性索引 + 需要回表（即该索引生成的计划是 IndexLookupReader）”时，选择满足条件的回表行数最小的索引作为候选索引。
```

File: choose-index.md [hunk 3]
```diff
@@ -29,19 +29,35 @@ aliases: ['/docs-cn/dev/choose-index/']
...
+上述规则中的“索引全匹配”指每个索引列上均存在等值条件。在执行 `EXPLAIN FORMAT = 'verbose' ...` 语句时，如果前置规则匹配了某一索引， TiDB 会输出一条 NOTE 级别的 warning 提示该索引匹配了前置规则。
+
+#### Skyline-Pruning
+
+Skyline-Pruning 是一个针对索引的启发式过滤规则，能降低错误估算导致选错索引的概率。 Skyline-Pruning 会从以下三个维度衡量一个索引的好坏：
 
 - 索引的列涵盖了多少访问条件。“访问条件”指的是可以转化为某列范围的 `where` 条件，如果某个索引的列集合涵盖的访问条件越多，那么它在这个维度上更优。
 
-对于这三种维度，如果某个索引 `idx_a` 在**三个维度上都不比 `idx_b` 差**，且**有一个维度比 `idx_b` 好**，那么就会优先选择 `idx_a`。
+- 选择该索引读表时，是否需要回表（即该索引生成的计划是 IndexReader 还是 IndexLookupReader）。不用回表的索引在这个维度上优于需要回表的索引。如果均需要回表，则比较索引的列涵盖了多少过滤条件。“过滤条件”指的是可以根据索引判断的 `where` 条件。如果某个索引的列集合涵盖的访问条件越多，则回表数量越少，那么它在这个维度上越优。
```
Comment (qiancai) [inline choose-index.md:56]:
```suggestion
- 选择该索引读表时，是否需要回表（即该索引生成的计划是 IndexReader 还是 IndexLookupReader）。不用回表的索引在这个维度上优于需要回表的索引。如果均需要回表，则比较索引的列涵盖了多少过滤条件。过滤条件指的是可以根据索引判断的 `where` 条件。如果某个索引的列集合涵盖的访问条件越多，则回表数量越少，那么它在这个维度上越优。
```

**Review by qiancai — COMMENTED** [review]
@xuyifangreeneyes PTAL

File: choose-index.md (hunk 3 above)
Comment (xuyifangreeneyes) [reply choose-index.md:56]: 上一条中的“访问条件”需要去掉双引号吗

File: choose-index.md [hunk 4]
```diff
@@ -29,19 +29,35 @@ aliases: ['/docs-cn/dev/choose-index/']
...
+上述规则中的“索引全匹配”指每个索引列上均存在等值条件。在执行 `EXPLAIN FORMAT = 'verbose' ...` 语句时，如果前置规则匹配了某一索引， TiDB 会输出一条 NOTE 级别的 warning 提示该索引匹配了前置规则。
+
+#### Skyline-Pruning
+
+Skyline-Pruning 是一个针对索引的启发式过滤规则，能降低错误估算导致选错索引的概率。 Skyline-Pruning 会从以下三个维度衡量一个索引的好坏：
 
 - 索引的列涵盖了多少访问条件。“访问条件”指的是可以转化为某列范围的 `where` 条件，如果某个索引的列集合涵盖的访问条件越多，那么它在这个维度上更优。
 
-对于这三种维度，如果某个索引 `idx_a` 在**三个维度上都不比 `idx_b` 差**，且**有一个维度比 `idx_b` 好**，那么就会优先选择 `idx_a`。
+- 选择该索引读表时，是否需要回表（即该索引生成的计划是 IndexReader 还是 IndexLookupReader）。不用回表的索引在这个维度上优于需要回表的索引。如果均需要回表，则比较索引的列涵盖了多少过滤条件。过滤条件指的是可以根据索引判断的 `where` 条件。如果某个索引的列集合涵盖的访问条件越多，则回表数量越少，那么它在这个维度上越优。
```
Comment (xuyifangreeneyes) [inline choose-index.md:56]: “访问条件”的双引号需要去掉吗

**Review by winoros — COMMENTED** [review]
We add two examples for the warning message output.
rest LGTM

**Comment by TomShawn** [comment]
> We add two examples for the warning message output.
> rest LGTM

So what changes should be made to this doc PR?

**Comment by xuyifangreeneyes** [comment]
> > We add two examples for the warning message output.
> > rest LGTM
> 
> So what changes should be made to this doc PR?

I have added the two examples in this PR, please take a look.

**Review by winoros — APPROVED** [review]

**Comment by qiancai** [comment]
@Rick-lee01 PTAL

File: choose-index.md [hunk 5]
```diff
@@ -29,19 +29,79 @@ aliases: ['/docs-cn/dev/choose-index/']
 
 ## 索引的选择
 
-TiDB 在选择索引时，会基于每个读表算子的代价估算，在此基础上提供了启发式规则 "Skyline-Pruning"，以降低错误估算导致选错索引的概率。
+TiDB 有基于规则和基于代价两种索引选择的方式。基于规则的索引选择包括前置规则和 Skyline-Pruning。在选择索引时，TiDB 会先尝试前置规则。如果存在索引满足某一条前置规则，则直接选择该索引。否则，TiDB 会采用 Skyline-Pruning 排除不合适的索引，然后基于每个读表算子的代价估算，选择代价最小的索引。
```
Comment (TomShawn) [inline choose-index.md:32]:
```suggestion
TiDB 基于规则或基于代价来选择索引。基于的规则包括前置规则和 Skyline-Pruning。在选择索引时，TiDB 会先尝试前置规则。如果存在索引满足某一条前置规则，则直接选择该索引。否则，TiDB 会采用 Skyline-Pruning 来排除不合适的索引，然后基于每个读表算子的代价估算，选择代价最小的索引。
```

File: choose-index.md [hunk 6]
```diff
@@ -29,19 +29,79 @@ aliases: ['/docs-cn/dev/choose-index/']
...
+### 基于规则选择
 
-Skyline-Pruning 是一个针对索引的启发式过滤规则，评判一个索引的好坏需要从以下三个维度进行衡量：
+#### 前置规则
 
-- 选择该索引读表时，是否需要回表（即该索引生成的计划是 IndexReader 还是 IndexLookupReader）。不用回表的索引在这个维度上优于需要回表的索引。
+TiDB 采用如下的启发式前置规则来选择索引：
 
-- 选择该索引是否能满足一定的顺序。因为索引的读取可以保证某些列集合的顺序，所以满足查询要求顺序的索引在这个维度上优于不满足的索引。
+1. 存在索引满足“唯一性索引全匹配 + 不需要回表（即该索引生成的计划是 IndexReader）”时，直接选择该索引。
```
Comment (TomShawn) [inline choose-index.md:40]:
```suggestion
+ 规则 1：如果存在索引满足“唯一性索引全匹配 + 不需要回表（即该索引生成的计划是 IndexReader）”时，直接选择该索引。
```

**Review by Rick-lee01 — APPROVED** [review]
LGTM

File: choose-index.md [hunk 7]
```diff
@@ -29,19 +29,79 @@ aliases: ['/docs-cn/dev/choose-index/']
...
+
+2. 存在索引满足“唯一性索引 + 需要回表（即该索引生成的计划是 IndexLookupReader）”时，选择满足条件的回表行数最小的索引作为候选索引。
```
Comment (TomShawn) [inline choose-index.md:42]:
```suggestion
+ 规则 2：如果存在索引满足“唯一性索引 + 需要回表（即该索引生成的计划是 IndexLookupReader）”时，选择满足该条件且回表行数最小的索引作为候选索引。
```

File: choose-index.md [hunk 8]
```diff
@@ -29,19 +29,79 @@ aliases: ['/docs-cn/dev/choose-index/']
...
+
+3. 存在索引满足“普通索引不需要回表 + 读取行数小于一定阈值”时，选择满足条件的读取行数最小的索引作为候选索引。
```
Comment (TomShawn) [inline choose-index.md:44]: “一定阈值”是多少呢？是由什么参数变量控制的吗？
```suggestion
+ 规则 3：如果存在索引满足“普通索引不需要回表 + 读取行数小于一定阈值”时，选择满足该条件且读取行数最小的索引作为候选索引。
```

File: choose-index.md [hunk 9]
```diff
@@ -29,19 +29,79 @@ aliases: ['/docs-cn/dev/choose-index/']
...
+
+4. 如果规则 2 和 3 仅有一条选出候选索引，则选择该候选索引。如果规则 2 和 3 均选出候选索引，选择读取行数（读索引行数 + 回表行数）较小的索引。
```
Comment (TomShawn) [inline choose-index.md:46]:
```suggestion
+ 规则 4：如果规则 2 和 3 之中仅选出一条候选索引，则选择该候选索引。如果规则 2 和 3 均选出候选索引，则选择读取行数（读索引行数 + 回表行数）较小的索引。
```

File: choose-index.md [hunk 10]
```diff
@@ -29,19 +29,79 @@ aliases: ['/docs-cn/dev/choose-index/']
...
+
+上述规则中的“索引全匹配”指每个索引列上均存在等值条件。在执行 `EXPLAIN FORMAT = 'verbose' ...` 语句时，如果前置规则匹配了某一索引， TiDB 会输出一条 NOTE 级别的 warning 提示该索引匹配了前置规则。在以下示例中，因为索引 `idx_b` 满足规则 2 中“唯一性索引 + 需要回表”的条件，TiDB 选择索引 `idx_b` 作为访问路径，`SHOW WARNING` 给出了索引 `idx_b` 命中前置规则的提示。
```
Comment (TomShawn) [inline choose-index.md:48]:
```suggestion
上述规则中的“索引全匹配”指每个索引列上均存在等值条件。在执行 `EXPLAIN FORMAT = 'verbose' ...` 语句时，如果前置规则匹配了某一索引，TiDB 会输出一条 NOTE 级别的 warning 提示该索引匹配了前置规则。

在以下示例中，因为索引 `idx_b` 满足规则 2 中“唯一性索引 + 需要回表”的条件，TiDB 选择索引 `idx_b` 作为访问路径，`SHOW WARNING` 返回了索引 `idx_b` 命中前置规则的提示。
```

File: choose-index.md [hunk 11]
```diff
@@ -29,19 +29,79 @@ aliases: ['/docs-cn/dev/choose-index/']
...
+| Level | Code | Message                                                                                   |
++-------+------+-------------------------------------------------------------------------------------------+
+| Note  | 1105 | unique index idx_b of t is selected since the path only has point ranges with double scan |
++-------+------+-------------------------------------------------------------------------------------------+
+1 row in set (0.00 sec)
+```
+
+#### Skyline-Pruning
+
+Skyline-Pruning 是一个针对索引的启发式过滤规则，能降低错误估算导致选错索引的概率。 Skyline-Pruning 会从以下三个维度衡量一个索引的好坏：
```
Comment (TomShawn) [inline choose-index.md:73]:
```suggestion
Skyline-Pruning 是一个针对索引的启发式过滤规则，能降低错误估算导致选错索引的概率。Skyline-Pruning 从以下三个维度衡量一个索引的优劣：
```

File: choose-index.md [hunk 12]
```diff
@@ -29,19 +29,79 @@ aliases: ['/docs-cn/dev/choose-index/']
...
 
 - 索引的列涵盖了多少访问条件。“访问条件”指的是可以转化为某列范围的 `where` 条件，如果某个索引的列集合涵盖的访问条件越多，那么它在这个维度上更优。
 
-对于这三种维度，如果某个索引 `idx_a` 在**三个维度上都不比 `idx_b` 差**，且**有一个维度比 `idx_b` 好**，那么就会优先选择 `idx_a`。
+- 选择该索引读表时，是否需要回表（即该索引生成的计划是 IndexReader 还是 IndexLookupReader）。不用回表的索引在这个维度上优于需要回表的索引。如果均需要回表，则比较索引的列涵盖了多少过滤条件。过滤条件指的是可以根据索引判断的 `where` 条件。如果某个索引的列集合涵盖的访问条件越多，则回表数量越少，那么它在这个维度上越优。
+
+- 选择该索引是否能满足一定的顺序。因为索引的读取可以保证某些列集合的顺序，所以满足查询要求顺序的索引在这个维度上优于不满足的索引。
+
+对于这三种维度，如果某个索引 `idx_a` 在**三个维度上都不比 `idx_b` 差**，且**有一个维度比 `idx_b` 好**，那么就会优先选择 `idx_a`。在执行 `EXPLAIN FORMAT = 'verbose' ...` 语句时，如果 Skyline-Pruning 排除了某些索引， TiDB 会输出一条 NOTE 级别的 warning 提示哪些索引在 Skyline-Pruning 之后保留下来。在以下示例中，索引 `idx_b` 和 `idx_e` 均劣于 `idx_b_c`，因而被 Skyline-Pruning 排除，`SHOW WARNING` 显示了经过 Skyline-Pruning 后剩余的索引。
```
Comment (TomShawn) [inline choose-index.md:81]:
```suggestion
对于以上三种维度，如果索引 `idx_a` 在这三个维度上都不比 `idx_b` 差，且有一个维度比 `idx_b` 好，那么 TiDB 会优先选择 `idx_a`。在执行 `EXPLAIN FORMAT = 'verbose' ...` 语句时，如果 Skyline-Pruning 排除了某些索引，TiDB 会输出一条 NOTE 级别的 warning 提示哪些索引在 Skyline-Pruning 排除之后保留下来。

在以下示例中，索引 `idx_b` 和 `idx_e` 均劣于 `idx_b_c`，因而被 Skyline-Pruning 排除，`SHOW WARNING` 的返回结果显示了经过 Skyline-Pruning 后剩余的索引。
```

File: choose-index.md (hunk 8 above)
Comment (xuyifangreeneyes) [reply choose-index.md:44]: 没有。这个小于一定阈值解释起来有一些复杂，对于普通索引 idx1，如果存在一个满足规则2中“唯一性索引 + 需要回表”的索引 idx2，idx1 的访问条件是 idx2 的超集，则可以确定 idx1 的读取行数小于 idx2 的读取行数，根据 idx2 索引唯一性 + 全匹配的特点可知 idx2 的读取行数是有限的，由此推断出 idx1 的读取行数是有限的。需要把这个细节加到文档里解释吗

Comment (TomShawn) [reply choose-index.md:44]: 不需要了，谢谢~

**Review by TomShawn — APPROVED** [review]
LGTM

File: choose-index.md [hunk 13]
```diff
@@ -29,19 +29,83 @@ aliases: ['/docs-cn/dev/choose-index/']
...
+#### Skyline-Pruning
+
+Skyline-Pruning 是一个针对索引的启发式过滤规则，能降低错误估算导致选错索引的概率。Skyline-Pruning 从以下三个维度衡量一个索引的优劣：
 
 - 索引的列涵盖了多少访问条件。“访问条件”指的是可以转化为某列范围的 `where` 条件，如果某个索引的列集合涵盖的访问条件越多，那么它在这个维度上更优。
 
-对于这三种维度，如果某个索引 `idx_a` 在**三个维度上都不比 `idx_b` 差**，且**有一个维度比 `idx_b` 好**，那么就会优先选择 `idx_a`。
+- 选择该索引读表时，是否需要回表（即该索引生成的计划是 IndexReader 还是 IndexLookupReader）。不用回表的索引在这个维度上优于需要回表的索引。如果均需要回表，则比较索引的列涵盖了多少过滤条件。过滤条件指的是可以根据索引判断的 `where` 条件。如果某个索引的列集合涵盖的访问条件越多，则回表数量越少，那么它在这个维度上越优。
+
+- 选择该索引是否能满足一定的顺序。因为索引的读取可以保证某些列集合的顺序，所以满足查询要求顺序的索引在这个维度上优于不满足的索引。
```
Comment (TomShawn) [inline choose-index.md:81]:
```suggestion
+ 选择该索引是否能满足一定的顺序。因为索引的读取可以保证某些列集合的顺序，所以满足查询要求顺序的索引在这个维度上优于不满足的索引。
```

**Pull request merged**